src/NJsonSchema.CodeGeneration.TypeScript/TypeScriptTypeResolver.cs

```diff
@@ -99,12 +99,21 @@ private string Resolve(JsonSchema4 schema, string typeNameHint, bool addInterfac
                 var prefix = addInterfacePrefix &&
                     SupportsConstructorConversion(schema.AdditionalPropertiesSchema) &&
                     schema.AdditionalPropertiesSchema?.ActualSchema.Type.HasFlag(JsonObjectType.Object) == true ? "I" : "";
-                var valueType = prefix + ResolveDictionaryValueType(schema, "any", Settings.SchemaType);
-                return $"{{ [key: string] : {valueType}; }}";
+
+                var valueType = prefix + ResolveDictionaryValueType(schema, "any");
+
+                var defaultType = "string";
+                var keyType = Settings.TypeScriptVersion >= 2.1m ? prefix + ResolveDictionaryKeyType(schema, defaultType) : defaultType;
+                if (keyType != defaultType)
+                {
+                    return $"{{ [key in keyof typeof {keyType}] : {valueType}; }}";
```

Comment:
The indexer should really be optional, i.e. `[key in keyof typeof {keyType}]?`. Otherwise TS will complain if you create an empty object map, which is a valid dictionary.

Reply:
I think the ? is added at another place when the property is not required/nullable. I think this is not the place to handle that.

Reply:
How would one mark the index property as not required though?

Here is the class and generated code that is causing me issues:

```csharp
public class AutoChecklistStateModel
{
    public Dictionary<AutoChecklistItem, bool> ItemsCompleted { get; set; } = new Dictionary<AutoChecklistItem, bool>();
}
```

```typescript
export interface AutoChecklistStateModel {
    itemsCompleted?: { [key in keyof typeof AutoChecklistItem]: boolean; } | null;
}

export enum AutoChecklistItem {
    Unknown = 0,
    FeedStockInventory = 1,
    MedicationStockInventory = 2,
    LiceCount = 3,
    FeedSystemFiles = 4,
}
```

Reply:
Causing the following error:

![image](https://user-images.githubusercontent.com/594564/72784660-9affc300-3c29-11ea-9275-0117d82d13fd.png)

Reply:
Ah yes, you are right. And changing this should not be a breaking change, right? Can you create a PR? I hope adding this is correct, I’m not that into the TS type system to know all consequences.

Reply:
I will create a PR. Since the new TS type will be a more relaxed superset of the current type, I can't see a way for it to break anything :)

Reply:
https://github.com/RicoSuter/NJsonSchema/pull/1120